File: server/datasets/java_source-code/Multiple-F1/type3_structural_33_merge_sort.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution {
    public int[] mergeSort(int[] arr) {
        if (arr == null || arr.length <= 1) return arr;
        int mid = arr.length / 2;
        int[] l = mergeSort(Arrays.copyOfRange(arr, 0, mid));
        int[] r = mergeSort(Arrays.copyOfRange(arr, mid, arr.length));
        
        List<Integer> listL = new ArrayList<>();
        List<Integer> listR = new ArrayList<>();
        for (int x : l) listL.add(x);
        for (int x : r) listR.add(x);
        
        int[] res = new int[arr.length];
        int idx = 0;
        while (!listL.isEmpty() && !listR.isEmpty()) {
            if (listL.get(0) <= listR.get(0)) res[idx++] = listL.remove(0);
            else res[idx++] = listR.remove(0);
        }
        while (!listL.isEmpty()) res[idx++] = listL.remove(0);
        while (!listR.isEmpty()) res[idx++] = listR.remove(0);
        return res;
    }
}

File: server/datasets/python_source-code/Multiple-F1/type3_structural_33_merge_sort.py
def merge_sort(arr):
    if len(arr) <= 1:
        return arr
    mid = len(arr) // 2
    l, r = merge_sort(arr[:mid]), merge_sort(arr[mid:])
    res = []
    while l and r:
        if l[0] <= r[0]:
            res.append(l.pop(0))
        else:
            res.append(r.pop(0))
    if l:
        res.extend(l)
    if r:
        res.extend(r)
    return res
